examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/user_data.c

```diff
@@ -0,0 +1,361 @@
+                                       USER_DATA_STR_SIZE / CRC_REGISTER_SIZE );
+        printf( "+ [0x%08lx] ", crc_code );
+    }
+    printf( "= [0x%08lx]", crc_code );
+    *out_checksum = crc_code;
+
+    /* DeInit */
+    if ( HAL_OK != HAL_CRC_DeInit( &crc_config ) )
+    {
+        printf( "\r\n\tCRC HAL Driver initialization [ERROR]" );
+        return -1;
+    }
+    __HAL_RCC_CRC_CLK_DISABLE();
+    return 0;
+}
+
+/**
+ * @brief  Calculate the CRC checksum of everything in *user_data but its CRC
+ *         checksum itself. If the CRC checksum we calculated matches the one
+ *         in the datastructure, return 0. Otherwise, return <0
+ * @param  *user_data is the structure whose integrity/usage is to be validated
+ * @retval 0: OK
+ * @retval -1: Failed to calculate checksum
+ * @retval -2: Calculated checksum doesn't match the one in user_data
+ */
+int8_t user_data_validate_checksum( user_data_t* user_data )
+{
+    int32_t new_checksum = 0x00000000;
+    if ( calculate_checksum( user_data, &new_checksum ) < 0 )
+    {
+        return -1;
+    }
+    if ( new_checksum != user_data->crc_checksum )
+    {
+        printf( "\r\n>> User data [CORRUPT] stored checksum %08lx doesn't match %08lx",
+                new_checksum, user_data->crc_checksum );
+        return -2;
+    }
+    printf( "\r\n>> User data integrity validation [OK]" );
+    return 0;
+}
+
+/******************************************************************************
+*                                     Helpers                                 *
+******************************************************************************/
+/**
+ * @brief Print a debug message displaying the contents of user_data, both in
+ *        binary format and ASCII.
+ *        A memory dump of the data itself is performed as to make it obvious to
+ *        the developer whether there was anything stored in flash, help debug
+ *        issues when modifying the user_data_t datastructure, etc.
+ * @param user_data is the structure we'd like to print to UART
+ */
+void user_data_printf( user_data_t* user_data )
+{
+#if 0
+    /* This debug message dumps the given *user_data memory block. Very useful
+    to find possible errors when modifying the user_data_t datastructure */
+    for ( uint32_t i = 0; i < sizeof( user_data_t ) / sizeof( int32_t ); i++ )
+    {
+        if ( 0 == i % 4 )
+        {
+            printf( "\r\n\t\t" );
+        }
+        printf( "%08lx ", *( ( int32_t* )user_data + i ) );
+    }
+#endif
+    printf( "\r\n\t* WiFi Security: [%ld]", user_data->wifi_client_encryption_mode );
+    printf( "\r\n\t* WiFi SSID: [%.64s]", user_data->wifi_client_ssid );
+    printf( "\r\n\t* WiFi Pwd: [%.64s]", user_data->wifi_client_password );
+    printf( "\r\n\t* Xi Acc ID: [%.64s]", user_data->xi_account_id );
+    printf( "\r\n\t* Xi Dev ID: [%.64s]", user_data->xi_device_id );
+    printf( "\r\n\t* Xi Dev Pwd: [%.64s]", user_data->xi_device_password );
```

Comment:
I think we shouldn't print out the xively device password to serial.  It would comment out this line and add a note saying that they COULD enable this if they need to debug something, but it's suppressed for security reasons.

Perhaps also fro the Wifi PWD, too.